_studio/shared/include/mfx_utils.h

```diff
@@ -145,6 +145,12 @@ constexpr const T& clamp( const T& v, const T& lo, const T& hi, Compare comp )
 {
     return comp(v, lo) ? lo : comp(hi, v) ? hi : v;
 }
+
+template<class T> inline
+T align_value(size_t value, size_t alignment = 16)
+{
```

Comment:
This will only work for `[alignment = power of two]` and fail for, as e.g.:
`x = align_value(2, 3); //expects 3 but result is 4`.
IMO, we need to add assertion that `alignment % 2 == 0`.
Yet another option is
```
size_t align(size_t value, size_t alignment)
{
    return alignment ?
        alignment * ((value + alignment - 1) / alignment) : value;
};
```
but it looks overkill on our case (MSDK doesn't use such weird alignments)

The second thing is using of arg. type deduction instead of specifying it explicitly -
```
template <typename T>
inline T align_value(T value, size_t alignment = 16);
```
since common usage pattern is 
`value  = mfx::align_value<type_of_value>(value,  alignment);`

Using explicit specification is error-prone, consider this
`value  = mfx::align_value<int16>(value,  alignment);` when actual type of value is `int32`

Reply:
@dmitry-gurulev good catch!
actually, ```alignment % 2 == 0``` doesn't mean that ```alignment``` is power of two :)
I think we can add in if statement to switch between realization to avoid division (actually need to benchmark and figure out if it has any sense to introduce complicated logic at all)

Agree with you about type deduction.

So currently I'm thinking about:
```
template<class T> inline
constexpr T align_value(size_t value, size_t alignment = 16)
{
   if (is_power_of_2(alignment))
      return static_cast<T> ((value + (alignment - 1)) & ~(alignment - 1));

   return alignment ?
        alignment * ((value + alignment - 1) / alignment) : value;
}
```

Reply:
@mgonchar,
> actually, alignment % 2 == 0 doesn't mean that alignment is power of two :)

Yes, u're right =) Maybe this one - `assert((alignment & (~alignment + 1)) == alignment);` ?

>I think we can add in if statement to switch between realization to avoid division

I don't think the division is show stopper, compilers are smart enough to replace it w/ shift when `alignment` is compile time constant, for me branching is worse. Agree that benchmark is required rather than speculations though.

Anyway, I still think this is overkill for MSDK, we don't use such alignments and we aren't common purpose lib like STL.

Reply:
@dmitry-gurulev If we decide to drop support of non-power of two alignments (which is OK as for me), I think we should change the name to reflect it somehow. Like ```align_value_to_power_of_2```.

Yes, your variant of power of two checking should work) But it will say true for zero, not sure if it is an actual problem

Reply:
align2_value